Orders: Display split view on large screens

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #6381 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR updates the Orders tab to show split view on large screens to take advantage of the available real estate.

The solution includes:
- Add a custom split view controller to be reused within the app. This split view controller uses the modern API to configure a double-column layout with the primary column being displayed when collapsed.
- Update the root view of the Orders tab to a new view controller that wraps the old order root within a split view controller. This is needed because the modern split view controller (initiated with `style`) doesn't work properly when added directly to a tab view controller.
- Extract `GhostTableViewController` to be a reusable type that displays ghost animation over a table view.
- Set up the orders split view controller with the order root and a ghost table view controller.
- Update the order list to automatically update the split view's detail view after loading orders
- Replace the order detail storyboard with a nib file and update the order detail view controller's initializer for easy setup.
- Additionally: update constraints on order root to make the view expand edge-to-edge in landscape mode.
- Also: updated UI tests to reflect the change to the Orders tab

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
On iPad (or large iPhones in landscape mode):
- Log in to a valid account. Notice that the details for the Orders tab in the tab bar are correct.
- Select the Orders tab. Notice that the list is displayed as a split view, with a placeholder view for the 2nd column initially.
- After a while, notice that details for the first item in the list are displayed on the 2nd column.
- Select another item, notice that the correct details are displayed on the 2nd column.

On iPhone (with small size):
- Log in to a valid account. Notice that the details for the Orders tab in the tab bar are correct.
- Select the Orders tab. Notice that the order list is displayed correctly.
- Select any order, notice that the correct details are displayed.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
On iPad:
https://user-images.githubusercontent.com/5533851/157225577-8f6aa477-867a-452c-90d6-665449408108.mp4

On iPhone:
https://user-images.githubusercontent.com/5533851/157227051-67d31080-cb80-4693-b886-c491c27bba19.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Replace Orders storyboard with OrderDetailsViewController xib

itsmeichigo committed:
Add split view for orders

itsmeichigo committed:
Update order detail when selecting new order

itsmeichigo committed:
Set preferred display mode for split view

itsmeichigo committed:
Make orders screen extend edge-to-edge

itsmeichigo committed:
Embed order detail in navigation controller

itsmeichigo committed:
Add ghost table view controller to display placeholder view

itsmeichigo committed:
Cleanup set up for orders split view

itsmeichigo committed:
Programmatically select first row in order list

itsmeichigo committed:
Create view controller to wrap split view

itsmeichigo committed:
Add custom split view controller to be shared within the app

itsmeichigo committed:
Only show details for first item if the order details screen is not displayed already

itsmeichigo committed:
Update release notes

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-split-view-orders

itsmeichigo committed:
Scroll content of details view to top before disappearing

itsmeichigo committed:
Update unit tests for main tab bar controller

itsmeichigo committed:
Fix issue checking for static text visibility in XCUIElement

itsmeichigo committed:
Fix failed UI tests for order list

itsmeichigo committed:
Fix failed UI tests for order details

itsmeichigo committed:
Fix typos

itsmeichigo committed:
Remove check for order details navigation bar (for now)

itsmeichigo committed:
Update test to check navigation bar title on order detail screen

pmusolino commented on RELEASE-NOTES.txt:
```diff
@@ -2,6 +2,7 @@
 
 8.8
 -----
+- [*] Show split view for Orders tab on large screens. [https://github.com/woocommerce/woocommerce-ios/pull/6387]
```
small nit: since this functionality has a strong impact on Orders and on our users and should be tested deeply, maybe it deserves two stars.

itsmeichigo replied:
Good point! Updated in a5fbc69d51d63f046f0a375dbc416415dc6d521b.

pmusolino commented on WooCommerce/Classes/ViewRelated/MainTabBarController.swift:
```diff
@@ -448,7 +448,7 @@ private extension MainTabBarController {
 
     func createOrdersViewController(siteID: Int64) -> UIViewController {
-        OrdersRootViewController(siteID: siteID)
+        OrdersSplitViewWrapperController(siteID: siteID)
```
I wasn't able to test this change with Push Notifications, but it's something we should consider before merging this.

itsmeichigo replied:
Thank you for reminding me, presenting order details indeed has some issue when the split view is introduced. This has been fixed in 7a405044e4eb28acc074248b94f9db476a0fda98.

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift:
```diff
@@ -48,22 +52,18 @@ final class OrderDetailsViewController: UIViewController {
-            reloadTableViewSectionsAndData()
-        }
-    }
+    private let viewModel: OrderDetailsViewModel
```
Nice refactoring!

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift:
```diff
@@ -93,6 +93,13 @@ final class OrderDetailsViewController: UIViewController {
+        // shows the detail screen, the table view content is at the top position
+        tableView.contentOffset = .zero
+        super.viewWillDisappear(animated)
+    }
```
If I understood the meaning of what you want to achieve here, I believe there is an error and probably bad behavior for the end user.
In theory, after selecting a new order, I should always see the beginning of that order. Instead, the offset is set to the same offset as the previous order. It could be ok for the user, but I believe that setting the `contentOffset` to zero here is creating the behavior that you can see below in the video. Basically, after switching tab, the view will scroll on top with some delay, but there is also a margin on top (it's not really offset zero).
Plus, you can notice that the selected order in the list, is no more selected after switching to a different tab.

https://user-images.githubusercontent.com/495617/157273616-255c0583-299a-4bbf-904d-8f1038a07f40.mp4

itsmeichigo replied:
I was reusing the order details screen for every selected order, so I wanted to scroll the details view to the top every time a new item is selected, so that it looks better in collapsed mode.

However, I haven't yet found a good way to scroll the table view to top successfully, so I switched to creating a new details view for every item in 47d7489ea650f57e5443c9821b458e0e71efa684. This will keep the old behavior that we had with the navigation stack.

> Plus, you can notice that the selected order in the list, is no more selected after switching to a different tab.

This is actually a different issue was handled in 3132df8a4de80b33efac478d545ba73d14c9224f and 680648fa38288ccd0a0d4be490858db27b4fe618.

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift:
```diff
@@ -123,6 +140,7 @@ private extension OrderDetailsViewController {
         tableView.rowHeight = UITableView.automaticDimension
 
         tableView.dataSource = viewModel.dataSource
+        tableView.accessibilityIdentifier = "order-details-table-view"
```
💯

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -350,7 +358,12 @@ extension OrderListViewController: SyncingCoordinatorDelegate {
         let headerContainer = UIView(frame: CGRect(x: 0, y: 0, width: Int(tableView.frame.width), height: 0))
         headerContainer.addSubview(topBannerView)
-        headerContainer.pinSubviewToSafeArea(topBannerView)
+        NSLayoutConstraint.activate([
+            headerContainer.leadingAnchor.constraint(equalTo: topBannerView.leadingAnchor),
+            headerContainer.trailingAnchor.constraint(equalTo: topBannerView.trailingAnchor),
+            headerContainer.safeAreaLayoutGuide.topAnchor.constraint(equalTo: topBannerView.topAnchor),
+            headerContainer.safeAreaLayoutGuide.bottomAnchor.constraint(equalTo: topBannerView.bottomAnchor),
+        ])
```
Q: It's not clear to me why this change. Why we don't want to maintain leading and trailing anchors to the safe area?

itsmeichigo replied:
I want the Order list to extend to all edges - this should be safe to do when all its subviews are anchored to safe area. I fixed the issue with the layout of the footer spinner view as well as the filters bar that you captured above in 09895c0523eb83d54bcff15cf709c7f49b84658a.

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -394,6 +407,25 @@ extension OrderListViewController {
+            return
+        }
+        let firstIndexPath = IndexPath(row: 0, section: 0)
+        tableView.selectRow(at: firstIndexPath, animated: false, scrollPosition: .top)
```
nit: maybe this is superfluous since we are already calling below `tableView(tableView, didSelectRowAt: firstIndexPath)`?

itsmeichigo replied:
The `selectRow` method only highlight the specified row and doesn't trigger the delegate method, therefore I need both methods to both highlight the row and trigger the selection.

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -515,7 +547,9 @@ private extension OrderListViewController {
 
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
-        tableView.deselectRow(at: indexPath, animated: true)
+        if splitViewController?.isCollapsed == true {
+            tableView.deselectRow(at: indexPath, animated: true)
+        }
```
I noticed that the cell remain selected after rotating back in portrait:
<img src="https://user-images.githubusercontent.com/495617/157279549-73103f39-3aec-4558-af52-8c6ba391c6d7.png" width=300 />

itsmeichigo replied:
This has been fixed in 3132df8a4de80b33efac478d545ba73d14c9224f by updating the highlight state on `viewDidLayoutSubview`.

pmusolino commented on WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift:
```diff
@@ -61,8 +68,9 @@ final class OrdersRootViewController: UIViewController {
         super.init(nibName: Self.nibName, bundle: nil)
 
+        // workaround to get rid of the extra space at the bottom when embedded in split view
+        extendedLayoutIncludesOpaqueBars = true
```
🤭 I think optimizing for big screens will give us a big headache. It is full of small details that are easy to forget.

itsmeichigo replied:
I believe we use opaque bars somewhere in the app (the navigation bars maybe), which causes the need for this. The default behavior for bars since iOS 7 is translucent, and this property only needs to be set to true if we use opaque bars.

itsmeichigo replied:
I can confirm that this only happens with views embedded in a navigation controller that is added to a split view. Since we're using opaque navigation bars, `extendedLayoutIncludesOpaqueBars` needs to be true to get rid of the extra space at the bottom. ~Removed redundant settings for navigation controllers in d9b192ab8a0de2e463ca02b0c7b051507d1716e0.~ (reverted the change because it doesn't work on small screens 😓)

itsmeichigo replied:
Come to think of it, maybe we should make sure that all child view controllers of our custom navigation bar has `extendedLayoutIncludesOpaqueBars` set to true. Updated this in 8a6cc081f63f95531de7edc63466ae8b792924b7.

pmusolino commented on WooCommerce/UITestsFoundation/Screens/Orders/OrdersScreen.swift:
```diff
@@ -48,10 +48,11 @@ public final class OrdersScreen: ScreenObject {
-        app.assertTextVisibilityCount(textToFind: String(orders[0].id))
-        app.assertTextVisibilityCount(textToFind: String(orders[0].total))
-        app.assertLabelContains(firstSubstring: String(orders[0].id), secondSubstring: orders[0].billing.first_name)
+        let ordersTableView = app.tables.matching(identifier: "orders-table-view")
+        XCTAssertEqual(orders.count, ordersTableView.cells.count, "Expecting '\(orders.count)' orders, got '\(app.tables.cells.count)' instead!")
+        ordersTableView.element.assertTextVisibilityCount(textToFind: String(orders[0].id))
+        ordersTableView.element.assertTextVisibilityCount(textToFind: String(orders[0].total))
+        ordersTableView.element.assertLabelContains(firstSubstring: String(orders[0].id), secondSubstring: orders[0].billing.first_name)
```
Thanks for updating our UI Tests!

pmusolino reviewed (changes requested):
Nice work @itsmeichigo! This is the first for a big refactoring of the app for supporting large screens and given the changes of this PR... this scares me a little bit 🤣 

By the way, I left some comments and I found some issues that should be addressed before merging the PR 👍 

<img src="https://user-images.githubusercontent.com/495617/157281861-79cd433e-528e-4174-8c92-cc562532337f.png" width=300 />

<img src="https://user-images.githubusercontent.com/495617/157282107-ef5a3d0b-c96c-499e-843a-fc76ac344d88.png" width=300 />

<img src="https://user-images.githubusercontent.com/495617/157282017-ffcb46c9-c231-45d5-b7fb-51dd621840f8.png" width=300 />

<img src="https://user-images.githubusercontent.com/495617/157282210-559f4330-8f5a-4e61-854e-58e77c050b32.png" width=300 />

<img src="https://user-images.githubusercontent.com/495617/157282257-078e8a75-4d71-42ee-b3f5-617d3d2bab96.png" width=300 />

itsmeichigo committed:
Update content of filters bar and footer spinner view to safe area

itsmeichigo committed:
Increase importance for the update in release notes

itsmeichigo committed:
Update navigation bar title of order details when switching items

itsmeichigo committed:
Update table header height on order list when layout subviews

itsmeichigo committed:
Stop reusing orderDetailsNavigationController

itsmeichigo committed:
Update highlight row on order list properly

itsmeichigo committed:
Fix issue presenting details for order notifications

itsmeichigo committed:
Present the navigation controller instead of the loader view controller directly

itsmeichigo committed:
Update highlight state after table reload due to snapshot change

itsmeichigo committed:
Highlight row after selecting programmatically

itsmeichigo commented:
Thank you @pmusolino for the thorough review! I've added a few more changes to fix issues with the layout and handling notifications. The PR is ready for another review 🙇‍♀️

itsmeichigo commented:
> <img src="https://user-images.githubusercontent.com/5533851/157414379-d680c505-0faa-4b0c-8312-5eadbb3e7d17.png" width=500 />

Regarding the issue with the color for the selected row, I'd need an opinion from @woocommerce/mobile-designers. By default when a row is selected, its separators are hidden. We may need to add custom views for the separators / or update the highlight color to something that stands out more.

itsmeichigo committed:
Remove redundant custom `extendedLayoutIncludesOpaqueBars` for navigation controllers

itsmeichigo committed:
Fix white space violation

itsmeichigo committed:
Revert "Fix white space violation"

This reverts commit 782bb34100e03a89b80c95c7cca2f741fe7b91f1.

itsmeichigo committed:
Revert "Remove redundant custom `extendedLayoutIncludesOpaqueBars` for navigation controllers"

This reverts commit d9b192ab8a0de2e463ca02b0c7b051507d1716e0.

itsmeichigo committed:
Display offline banner on the split view wrapper instead of the child view controllers

itsmeichigo committed:
Make sure that all child view controllers of WooNavigationController has extendedLayoutIncludesOpaqueBars set to true

itsmeichigo committed:
Merge branch 'issue/6381-split-view-orders' into issue/6381-split-view-orders-enhancements

itsmeichigo committed:
Set empty view for the detail view of orders split view

joe-keenan commented:
I’ve looked around at different approaches to this, and I think the simplest solution, for now, is to not hide the separators.

This would ideally apply to both tap and selected states – because at the moment, if you tap on a single row on a grey background, the whole row looks hidden. If we keep the separators, then the row will still look like a row, even though the background colours are the same.

This will have the same effect on this split-screen view.

Current on-tap look:
![image](https://user-images.githubusercontent.com/95003933/157601995-3bb1f5bf-0038-43ee-bead-bf7c4bcf9fad.png)

Revised look with separators
![image](https://user-images.githubusercontent.com/95003933/157602666-a88b6f4e-ac02-4464-b2c0-dcd9a63c8e78.png)


Does this sound ok?

pmusolino reviewed (commented):
I re-tried this PR, but while I was testing it on an iPad Pro (12.9 inch) 5th generation with iOS 15.2, I encountered a crash.
This is what I did:
1. I moved to Orders, in landscape mode.
2. I loaded all the orders (paginated). 
3. I returned to top, and I did more than a pull to refresh.
4. Crash 💥 

The state of the view became strange: the data on the left were no longer displayed, while on the right the ghost view disappear displaying one order.
![Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2022-03-10 at 11 39 53](https://user-images.githubusercontent.com/495617/157645571-c745ce46-88d4-4bfd-832c-604b401d7d15.png)

At the same time, the crash happened.
<img width="1247" alt="Schermata 2022-03-10 alle 11 40 12" src="https://user-images.githubusercontent.com/495617/157645644-5b0686e1-e522-49c8-9bf3-faa6c8a0058b.png">

itsmeichigo commented:
Hi @pmusolino - can you reproduce the issue repeatedly? I'm wondering if this is an existing issue of the order list, because I haven't touched the data part of the list.

pmusolino commented:
> Hi @pmusolino - can you reproduce the issue repeatedly? I'm wondering if this is an existing issue of the order list, because I haven't touched the data part of the list.

Yes, I'm able to reproduce it, but just on this PR. If I pull `trunk`, I'm not able to replicate the crash 🤔

itsmeichigo committed:
Only highlight row before displaying cells

itsmeichigo committed:
Add separators to the selected background view of order cells

itsmeichigo committed:
Update constraints for the bottom separator

itsmeichigo committed:
Update separator height

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-split-view-orders

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-split-view-orders

itsmeichigo committed:
Merge pull request #6399 from woocommerce/issue/6381-split-view-orders-enhancements

Orders: Enhancements to split view

itsmeichigo committed:
Restore logic in Order Roots view prior to split view refactor

itsmeichigo committed:
Add feature flags for the split view in MainTabBarController

itsmeichigo committed:
Add feature flag to the order list view controller

itsmeichigo committed:
Merge branch 'issue/6381-split-view-orders' into issue/6381-apply-feature-flag

itsmeichigo committed:
Switch the root controller for orders tab based on the feature flag

itsmeichigo committed:
Push the detail controller when selecting order

itsmeichigo committed:
Reload table view on rotation to update the selected state for the list in split view

itsmeichigo committed:
Update offline banner on order details and order root based on feature flag

itsmeichigo committed:
Hide updates to extendedLayoutIncludesOpaqueBars under feature flag

itsmeichigo committed:
Fix failed unit tests

itsmeichigo commented:
Hi @pmusolino! The crash seems to happen when we try to select a table view row when the table view is still loading from the snapshot updates. It might be safer to only update the selection state right before displaying the cells, so I move the highlight logic to the delegate method `willDisplayCell` in cb8c5b23a0b4bb4e5cee35f21e03ba5d39273c83. I believe it's safer there since we're dealing with solely the main thread in this case.

I've also added separators to the selected background view of order cells like @joe-keenan suggested:

![Simulator Screen Shot - iPhone 13 Pro Max - 2022-03-11 at 13 14 03](https://user-images.githubusercontent.com/5533851/157812800-78a150a2-5cdf-4136-85dd-ab35663dc980.png)

I've also added a PR to hide the changes in this PR under a feature flag: #6415. Please feel free to test that PR first!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift:
```diff
@@ -122,8 +122,35 @@ private extension OrderTableViewCell {
+            bottomSeparatorView.translatesAutoresizingMaskIntoConstraints = false
+            view.addSubview(bottomSeparatorView)
+            NSLayoutConstraint.activate([
+                view.bottomAnchor.constraint(equalTo: bottomSeparatorView.bottomAnchor, constant: separatorHeight/2),
+                view.leadingAnchor.constraint(equalTo: bottomSeparatorView.leadingAnchor),
+                view.trailingAnchor.constraint(equalTo: bottomSeparatorView.trailingAnchor),
+                bottomSeparatorView.heightAnchor.constraint(equalToConstant: separatorHeight/2)
```
I had to reduce the bottom separator height in half - this is hacky but it helps with making it look equal with other separators. Please let me know if there's a better solution!

itsmeichigo committed:
Remove split view from release notes

itsmeichigo committed:
Inject feature flag service to MainTabBarController

itsmeichigo committed:
Fix flaky tests by injecting feature flag service

pmusolino reviewed (approved):
Thanks for fixing all the issues @itsmeichigo 🚀

itsmeichigo committed:
Merge pull request #6415 from woocommerce/issue/6381-apply-feature-flag

Orders: Hide update of split view under feature flag

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-split-view-orders
